File: threedollar-api-user-service/src/integrationTest/java/com/depromeet/threedollar/api/user/service/auth/AuthServiceFinderTest.java
package com.depromeet.threedollar.api.user.service.auth;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.stream.Stream;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import com.depromeet.threedollar.domain.rds.domain.userservice.user.UserSocialType;

@SpringBootTest
class AuthServiceFinderTest {

    @Autowired
    private AuthServiceFinder authServiceFinder;

    @MethodSource("userSocialTypeAndServices")
    @ParameterizedTest
    void 각_소셜의_AuthService_구현체를_가져옵니다(UserSocialType socialType, Class<AuthService> expectedAuthServiceClass) {
        // when
        AuthService authService = authServiceFinder.getAuthService(socialType);

        // then
        assertThat(authService).isInstanceOf(expectedAuthServiceClass);
    }

    private static Stream<Arguments> userSocialTypeAndServices() {
        return Stream.of(
            Arguments.of(UserSocialType.KAKAO, KaKaoAuthService.class),
            Arguments.of(UserSocialType.APPLE, AppleAuthService.class),
            Arguments.of(UserSocialType.GOOGLE, GoogleAuthService.class)
        );
    }

}

File: threedollar-api-boss/src/integrationTest/kotlin/com/depromeet/threedollar/api/boss/service/auth/AuthServiceFinderTest.kt
package com.depromeet.threedollar.api.boss.service.auth

import java.util.stream.Stream
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.TestConstructor
import com.depromeet.threedollar.common.exception.model.ServiceUnAvailableException
import com.depromeet.threedollar.domain.mongo.boss.domain.account.BossAccountSocialType

@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
@SpringBootTest
internal class AuthServiceFinderTest(
    private val authServiceFinder: AuthServiceFinder
) {

    @MethodSource("bossSocialTypeAndServices")
    @ParameterizedTest
    fun `각 소셜타입의 AuthService를 가져온다`(socialType: BossAccountSocialType, expectedAuthServiceClass: Class<AuthService>) {
        // when
        val authService = authServiceFinder.getAuthService(socialType)

        // then
        assertThat(authService).isInstanceOf(expectedAuthServiceClass)
    }

    companion object {
        @JvmStatic
        private fun bossSocialTypeAndServices(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(BossAccountSocialType.KAKAO, KaKaoAuthService::class.java),
                Arguments.of(BossAccountSocialType.APPLE, AppleAuthService::class.java),
                Arguments.of(BossAccountSocialType.GOOGLE, GoogleAuthService::class.java)
            )
        }
    }

    @Test
    fun `네이버 인증방식은 아직 지원하지 않는다`() {
        // given
        val socialType = BossAccountSocialType.NAVER

        // when & then
        assertThatThrownBy { authServiceFinder.getAuthService(socialType) }.isInstanceOf(ServiceUnAvailableException::class.java)
    }

}
